### PR TITLE
Make `[T]::len` and `str::len` const fn

### PR DESCRIPTION
**Pull request by oli-obk**
r? @Gankro

### CODE REVIEW

**Comment by oli-obk** [comment]
cc @Centril

**Comment by leonardo-m** [comment]
A simple common use case:

```
fn main() {
    const A: [u32; 3] = [1, 2, 3];
    const B: [u32; A.len()] = [10, 20, 30];
}
```

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -119,9 +126,10 @@ impl<T> [T] {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
-    pub fn len(&self) -> usize {
+    #[rustc_const_unstable(feature = "const_slice_len")]
+    pub const fn len(&self) -> usize {
         unsafe {
-            mem::transmute::<&[T], Repr<T>>(self).len
```
Comment (Centril) [inline src/libcore/slice/mod.rs]: Nice!

**Comment by Gankro** [comment]
Seems solid. I don’t have review rights tho

**Comment by oli-obk** [comment]
Hot potato passing until success

r? @SimonSapin

**Comment by leonardo-m** [comment]
It doesn't work in this equally common case:

```
#![feature(const_slice_len)]
#![allow(unused_variables)]

fn main() {
    let a: [u8; 3] = [10, 20, 30];
    let b: [u8; a.len()] = Default::default();
    // error[E0435]: attempt to use a non-constant value in a constant
}
```

**Comment by oli-obk** [comment]
That will never work with const eval as it is setup currently. Supporting such cases (I'd call them "lazy const eval") would technically be possible but definitely requires an RFC.

File: src/libcore/str/mod.rs [hunk 2]
```diff
@@ -2242,8 +2244,15 @@ impl str {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline(always)]
-    pub fn as_bytes(&self) -> &[u8] {
-        unsafe { &*(self as *const str as *const [u8]) }
+    #[rustc_const_unstable(feature="const_str_as_bytes")]
+    pub const fn as_bytes(&self) -> &[u8] {
+        unsafe {
+            union Slices<'a> {
```
Comment (fbstj) [inline src/libcore/str/mod.rs]: can/should the union be moved outside the unsafe block? it's definition isn't unsafe iirc

Comment (oli-obk) [reply src/libcore/str/mod.rs]: While that could certainly be done, there's no difference here and this seems very self-contained to me. YMMV, so feel free to open a PR about that if you want :)

**Comment by rodrimati1992** [comment]
Is this likely to stabilize soon?I need either this or the layout of `&[T]` `str` to be guaranteed to be able to to construct ffi-safe slices in constants without using hacks.

**Comment by oli-obk** [comment]
Stabilization is blocked on https://github.com/rust-lang/rust/issues/51909

**Pull request merged**